OpenRA.Game/Traits/Target.cs

```diff
@@ -80,13 +80,20 @@ public TargetType Type
 
 		public bool IsValidFor(Actor targeter)
 		{
-			if (targeter == null || Type == TargetType.Invalid)
+			if (targeter == null)
 				return false;
 
-			if (actor != null && !actor.IsTargetableBy(targeter))
-				return false;
-
-			return true;
+			switch (Type)
+			{
+				case TargetType.Actor:
+					return actor.IsTargetableBy(targeter);
+				case TargetType.FrozenActor:
+					return frozen.IsValid && frozen.Visible && !frozen.Hidden;
+				case TargetType.Invalid:
+					return false;
+				default:
```

Comment:
Can we explicitly list `TargetType.Terrain` here instead of `default` to bring this in line with what we do everywhere else in the class and to make the code easier to understand?